cpp/src/arrow/compute/kernels/scalar_string.cc

```diff
@@ -427,6 +427,189 @@ void AddMatchSubstring(FunctionRegistry* registry) {
+          // write a single codepoint
+          PROPAGATE_FALSE(arrow::util::UTF8Decode(&i, &codepoint));
+          dest = arrow::util::UTF8Encode(dest, codepoint);
+          // and skip the remainder
+          int64_t skips = options.step - 1;
+          while ((skips--) && (i < end_sliced)) {
+            PROPAGATE_FALSE(arrow::util::UTF8Decode(&i, &codepoint));
+          }
+        }
+        *output_written = static_cast<offset_type>(dest - output);
+      }
+      return true;
+    } else {  // step < 0
+      // serious +1 -1 kung fu because now begin_slice and end_slice act like reverse
+      // iterators.
+
+      if (options.start >= 0) {
+        // +1 because begin_sliced acts as as the end of a reverse iterator
+        PROPAGATE_FALSE(arrow::util::UTF8AdvanceCodepoints(begin, end, &begin_sliced,
+                                                           options.start + 1));
+        // and make it point at the last codeunit of the previous codeunit
+        begin_sliced--;
+      } else {
+        // -1 because start=-1 means the last codeunit, which is 0 advances
+        PROPAGATE_FALSE(arrow::util::UTF8AdvanceCodepointsReverse(
+            begin, end, &begin_sliced, -options.start - 1));
+        // and make it point at the last codeunit of the previous codeunit
+        begin_sliced--;
+      }
+      // similar to options.start
+      if (options.stop >= 0) {
+        PROPAGATE_FALSE(arrow::util::UTF8AdvanceCodepoints(begin, end, &end_sliced,
+                                                           options.stop + 1));
+        end_sliced--;
+      } else {
+        PROPAGATE_FALSE(arrow::util::UTF8AdvanceCodepointsReverse(begin, end, &end_sliced,
+                                                                  -options.stop - 1));
+        end_sliced--;
+      }
+
+      uint8_t* dest = output;
+      const uint8_t* i = begin_sliced;
+
+      while (i > end_sliced) {
+        uint32_t codepoint = 0;
+        // write a single codepoint
+        PROPAGATE_FALSE(arrow::util::UTF8DecodeReverse(&i, &codepoint));
+        dest = arrow::util::UTF8Encode(dest, codepoint);
+        // and skip the remainder
+        int64_t skips = -options.step - 1;
+        while ((skips--) && (i > end_sliced)) {
+          PROPAGATE_FALSE(arrow::util::UTF8DecodeReverse(&i, &codepoint));
+        }
+      }
+      *output_written = static_cast<offset_type>(dest - output);
+      return true;
+    }
+  }
+};
+
+const FunctionDoc utf8_slice_codeunits_doc(
+    "Slice string ",
+    ("For each string in `strings`, slice into a substring defined by\n"
+     "`start`, `stop`, `step`) as given by `SliceOptions` where `start` is inclusive\n"
+     "and `stop` is exclusive and are measured in codeunits. If step is negative, the\n"
+     "string will be advanced in reversed order. A `step` of zero is considered an\n"
+     "error.\n"
+     "Null inputs emit null."),
+    {"strings"}, "SliceOptions");
+
+void AddSlice(FunctionRegistry* registry) {
```

Comment:
Is there a reason we don't also add a `ascii_slice` for binary and string?

Reply:
No, just focussing on the harder problems first.